File: src/ImageLoader.cpp
#include "ImageLoader.h"

#include <fstream>
#include <iostream>
#include <map>
#include <string>

using namespace std;

ImageLoader::ImageLoader(const std::string &path) 
    : path_(path) {
    load();
}

VImage ImageLoader::loadSingle(const std::string &path) {
    if(VIPS_INIT("ImageLoader"))
        throw std::runtime_error("Impossible d'initialiser libvips");

    VImage img = VImage::new_from_file(path.c_str(),
                                       VImage::option()->set("access","random"));

    width_ = img.width();
    height_ = img.height();

    // RGB to grayscale, only one channel holds the data
    VImage r = img[0];
    VImage g = img[1];
    VImage b = img[2];

    VImage gray = r.ifthenelse(r, g.ifthenelse(g, b));

    return gray;
}

void ImageLoader::load() {
    ifstream f(path_);
    if (!f.is_open()) {
        cerr << "Failed to open JSON file!" << endl;
        return;
    }
    data_ = json::parse(f);
    
    if (data_.contains("imgs") && data_["imgs"].is_object()) {
        for (auto& [key, pathJson] : data_["imgs"].items()) {
            string path = pathJson.get<string>();
            
            try {
                VImage img = loadSingle(path);
                imageMap_[key] = img;
                height_ = img.height();
                width_ = img.width();

                cout << "Loaded image for " << key << " from " << path << endl;
            } catch (const exception& e) {
                cerr << "Failed to load image " << key << ": " << e.what() << endl;
            }
        }
    } else {
        cerr << "\"imgs\" section missing or invalid in JSON" << endl;
    }

    // Add dummy black image
    imageMap_["0"] = VImage::black(width_, height_);
}

std::vector<std::tuple<int,int,VImage,bool>> ImageLoader::prepareViews() const {
    std::vector<std::tuple<int,int,VImage,bool>> views;

    if (!data_.contains("views") || !data_["views"].contains("datas"))
        return views;

    for (const auto& viewData : data_["views"]["datas"]) {
        int row = viewData.value("row", 0);
        int col = viewData.value("col", 0);
        bool grayscale = viewData.value("grayscale", true);

        VImage img;

        if (grayscale) {
            std::string chan = viewData.value("r", "0");
            img = imageMap_.at(chan);
        } else {
            std::string rChan = viewData.value("r", "0");
            std::string gChan = viewData.value("g", "0");
            std::string bChan = viewData.value("b", "0");

            img = VImage::bandjoin({
                imageMap_.at(rChan),
                imageMap_.at(gChan),
                imageMap_.at(bChan)
            });
        }

        views.emplace_back(row, col, img, grayscale);
    }

    return views;
}

File: viewer/imageLoader.py
import json
from rich.progress import Progress, SpinnerColumn, TextColumn, BarColumn, TimeElapsedColumn
from importlib import resources
import numpy as np
import pyvips

from viewer.logger import Logger

class ImageLoader:
    def __init__(self, path):
        self.path = path
        self.width = 0
        self.height = 0
        self.data = {}
        self.image_map = {}
        self.log = Logger("ImageLoader").logger
        self.load()

    def load_single(self, path):
        # Chargement avec pyvips
        img = pyvips.Image.new_from_file(path, access="random")

        self.width = img.width
        self.height = img.height

        r = img[0]
        g = img[1]
        b = img[2]

        gray = r.ifthenelse(r, g.ifthenelse(g, b))

        # Conversion en numpy et passage en niveau de gris
        arr = np.ndarray(buffer=gray.write_to_memory(),
                         dtype=np.uint8,
                         shape=(gray.height, gray.width))

        return arr

    def load(self):
        # Chargement du JSON
        try:
            with open(self.path, "r") as f:
                self.data = json.load(f)
        except Exception as e:
            self.log.error(f"Failed to open JSON file: {e}")
            return

        imgs = self.data.get("imgs", {})
        if not isinstance(imgs, dict):
            self.log.warning('"imgs" section missing or invalid in JSON')
            return

        with Progress(
            SpinnerColumn(),
            TextColumn("[progress.description]{task.description}"),
            BarColumn(),
            TextColumn("{task.completed}/{task.total}"),
            TimeElapsedColumn(),
        ) as progress:
            task = progress.add_task("Loading images...   ", total=len(imgs))

            for key, path in imgs.items():
                try:
                    img_path = resources.files("resources.data") / path
                    img = self.load_single(img_path)
                    self.image_map[key] = img
                    self.height, self.width = img.shape[:2]
                    self.log.info(f"Loaded image '{key}' from {path}")
                except Exception as e:
                    self.log.error(f"Failed to load image '{key}': {e}")
                finally:
                    progress.advance(task)

        self.image_map["0"] = np.zeros((self.height, self.width), dtype=np.uint8)

    def prepare_views(self):
        views = []

        views_data = self.data.get("views", {}).get("datas", [])
        if not views_data:
            self.log.warning("No views found in JSON")
            return views

        # Progress bar pour les vues
        with Progress(
            SpinnerColumn(),
            TextColumn("[progress.description]{task.description}"),
            BarColumn(),
            TextColumn("{task.completed}/{task.total}"),
            TimeElapsedColumn(),
        ) as progress:
            task = progress.add_task("Preparing views...  ", total=len(views_data))

            for view in views_data:
                row = view.get("row", 0)
                col = view.get("col", 0)
                grayscale = view.get("grayscale", True)

                if grayscale:
                    chan = view.get("r", "0")
                    img = self.image_map.get(chan, self.image_map["0"])
                else:
                    r_chan = view.get("r", "0")
                    g_chan = view.get("g", "0")
                    b_chan = view.get("b", "0")
                    # Assemblage RGB
                    img = np.stack([
                        self.image_map.get(r_chan, self.image_map["0"]),
                        self.image_map.get(g_chan, self.image_map["0"]),
                        self.image_map.get(b_chan, self.image_map["0"])
                    ], axis=2)

                views.append((row, col, img, grayscale))
                progress.advance(task)

        return views
